### PR TITLE
[feat] Reinforcement learning with GRPO and DPO

### LINKED ISSUE #8: Add support for GRPO and DPO trainer
GRPO Trainer and Config works similarly as SFTTrainer and SFTConfig format.

Read more here: https://huggingface.co/docs/trl/grpo_trainer

Here are few key notes I got from a first reading:

1. We need to update how data preprocessing is done separately for RL dataset to handle reasoning, CoT

2. We need to provide some very simple sample reward functions but since RL tasks can be very specific the user most likely will just upload a file consisting reward functions (we will check these using the specified function prototype requirement in the docs)

3. Need to setup vLLM ("colocate") mode since Cloud Run GPU only supports **ONE** GPU so we cannot run another for inference -- but other than that should just be installing `pip install trl[vllm]`

```python
training_args = GRPOConfig(
    ...,
    use_vllm=True,
    vllm_mode="colocate",
)
```

Would be good to test performance of this way of using vLLM to see if there is any gain at all...

In the future when we migrate this to Vertex AI pipeline consider using `vllm_mode="server"` but we need to run another separate inference process.

4. Most other setup are really similar like model / tokenizer setup, saving, PEFT, inference, etc.

Since this requires changes to preprocessing I will finish all other training stuff then come back to this.

### LINKED ISSUE #46: Extend reward functions support
We currently only provides a few very simple built-in reward functions for the user to select. To extend this to actual real world application, we should add:

1. using `AutoModelForSequenceClassification` (either string ID or `PreTrainedModel` object) as somewhat of a "reward model"
2. more built-in reward functions that are somwhat customizable, based on existing R1 and GMSK rewards, these can be toggles / separated by task by just returning None when not applicable
3. using LLM as judge in one of the custom reward functions

These are elaborated in the docs: https://huggingface.co/docs/trl/main/en/grpo_trainer#trl.GRPOTrainer.reward_funcs

Acceptance criteria: actually fine tune a gemma on a real world use case with these complex rewards

### PR DESCRIPTION
**Pull request by supreme-gg-gg**
## Summary

- Full refactoring of training config for modular parameter specification, see `jobs/training/Readme.md`
- Updates provider functions to setup GRPOConfig and GRPOTrainer same for DPO, closes #8 
- Update vision collator to only work with SFTTrainer and for RFT we reformat the vision dataset so the trainer will handle the image processing internally (this is the only way as per the TRL docs)
- Provided more reward functions / graders to choose from taking inspiration from OpenAI Fine tuning platform, OpenR1, and ART, including simple reward engineering and LLM-as-judge -- this is supported by a new interface for configuring rewards to use -- closes #46 
- Uses streaming in preprocessing service when sample size is specified to handle large datasets efficiently

## Note

Updated versions of TRL, transformers, and unsloth to work with GRPO. The user needs to provide Gemini API Key to access LLM-as-judge graders. The most recommended grader is `RULER`, need to provide simple custom rubric.

This has been tested on:

1. Text preference alignment with DPO (not vision, see below)
2. Text & reasoning with GRPO

The specific datasets and cookbook for reference are documented in detail in #8 

The specific design for the new rewards system is documented in `rewards.py` as well as #46 

The next step is #50, currently we use a workaround using the `_reformat_vision_dataset` function but since TRL updated its preferred format we shall sync up.

## Blocked

1. Using vLLM to accelerate GRPO training to make generation faster -- this is blocked since Unsloth does not yet support vLLM with Gemma3, I have not heard news about this in a bit will investigate and update when ready, but should be easy to update and use "colocate" mode

2. DPO vision with Gemma3, this is blocked because TRL DPOTrainer currently does not recognize Gemma3 as vision model and apply processor but instead applies tokenizer, opened an issue and will update when got news. There is a workaround using a custom data collator, but would prefer fixing the internal processing of the trainer 😄 

EDIT: It seems like workaround is not possible because the trainer first calls `_prepare_dataset` which creates this `ArrowInvalid: cannot mix list and non-list, non-null values` bug

3. GRPO is known to take forever in comparison to DPO and SFT, but Cloud Run Jobs with GPU has a limit of 3600 seconds before auto-terminating. The non-GPU option is very long but it's irrelevant. The only alternative, it seems, is to pivot to Vertex AI, but that's not in the scope of this issue 😭

### CODE REVIEW

**Comment by supreme-gg-gg** [comment]
@inclinedadarsh this is ready to be merged! I will ship the reward function enhancement in a separate PR because they're technically both standalone features and then ship the new frontend

**Comment by supreme-gg-gg** [comment]
@inclinedadarsh does this conflict with your work? If not I'll merge it, there are two limitations that are blocked for now but are in the PR descriptions, hope to ship them soon

**Comment by supreme-gg-gg** [comment]
For testing if you're doing multimodal / using LLM-as-judge graders, do not use max-step > 100, most of the times 30-40 steps will already cause job timeout (or try reducing batch size, gradient accumulation, no. of generation, seq length). However, GRPO usually work best if you train >300 steps, which under the current jobs config is only possible for simple built in deterministic rewards + text dataset

**Comment by inclinedadarsh** [comment]
This sounds good.

Good thing is, none of this conflicts with my work. feel free to merge it. I'm not sure if I'll be able to test it, but I'll try to go through the flow once.
